### PR TITLE
fix: fix optimizer step working on random initialized model weights

### PR DESCRIPTION
**Pull request by pengdurice**
## What does this PR do ?

Fix the bug of optimizing on random initialized model weights when a very very corner case combination is used when cpu offloading + distributed optimizer + precision = bf16 + using a hf checkpoint

## Issues
https://github.com/NVIDIA-NeMo/RL/issues/2371
run:
`NRL_DISABLE_FP32_SYNC_FIX=1 bash tests/test_suites/llm/sft-qwen2.5-0.5b-1n2g-megatron-cpuoffload.sh`
loss values ~10 which is logprobs of random prediction of ~100K vocab

Image

When calling with 
`NRL_DISABLE_FP32_SYNC_FIX=0 bash tests/test_suites/llm/sft-qwen2.5-0.5b-1n2g-megatron-cpuoffload.sh`

Image

## Note:
Some of the redundant debug codes can be cleaned up later, feel free to use them for reproducing the errors and fixes first.
## Usage
NA
## What is the bug
When cpu offload  + using bf16 + distributed optimizer is enabled and model is initialized from a checkpoints (i.e. fine tuning), a hidden bug manifests. The reason is the bf16 cpu copy and the fp32 cpu copy for the hybrid device optimizer are not correctly flushed by the imported model checkpoints, i.e. they remain random initialized values; and later then the optimizer's step() works on them.
## The fix
Force copying the hf checkpoint's model weights to these random initialized copies 
## Before your PR is "Ready for review"
**Pre checks**:
- [X] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

## Additional Information
* ...

### CODE REVIEW

**Comment by pengdurice** [comment]
@terrykong, Hi Terry, feel free to take a look at the issue (https://github.com/NVIDIA-NeMo/RL/issues/2371) and this PR when you get a chance!

**Review by terrykong — COMMENTED** [review]
@cuichenx @yaoyu-33 could you review?

**Comment by pengdurice** [comment]
@cuichenx @yaoyu-33 , hi nice to meet you folks, this is Peng. This PR is to resolve an optimizer working on random model weight issue. Would you please review it? thank you!

**Review by cuichenx — DISMISSED** [review]
Thanks — LGTM from a Bridge perspective. Same bug shape exists in Megatron-Bridge (HF → mcore init + optimizer_cpu_offload=True + distributed optimizer + BF16 hits the same stale FP32 master copies in HybridDeviceOptimizer), so I'll mirror this workaround on the Bridge side.

 The root cause looks like it's in mcore — reload_model_params() only walks level 1 of HybridDeviceOptimizer and leaves levels 2/3 (CPU clones, FP32 working copy) at random init. Might be worth filing an mcore issue so the fix lands upstream and both RL and Bridge can drop their copies of this helper eventually — the workaround reaches into a few private attrs (_get_model_param_range_map, gpu_params_map_cpu_copy, param_to_fp32_param), so it's a bit fragile to carry long-term.

**Comment by pengdurice** [comment]
Thanks for the review @cuichenx! This is approved now but the workflows are still awaiting maintainer approval ("10 workflows awaiting approval"). Could a maintainer approve the CI runs (or /ok to test)? thanks! cc @terrykong

File: nemo_rl/models/megatron/setup.py [hunk 1]
```diff
@@ -1196,6 +1242,12 @@ def composed_peft_hook(model: list[MegatronModule]) -> list[MegatronModule]:
             skip_load_to_model_and_opt=HAVE_FSDP2 and megatron_cfg.dist.use_torch_fsdp2,
         )
         print("Checkpoint loaded")
+
+        # See _force_sync_optimizer_fp32_from_model: required when
+        # optimizer_cpu_offload=True so the first optimizer step does Adam on the
+        # loaded HF weights instead of stale random init in the FP32 master copies.
+        if optimizer is not None:
```
Comment (RayenTian) [inline nemo_rl/models/megatron/setup.py]: One thing I wanted to double-check with you: the `_force_sync_optimizer_fp32_from_model` call sits inside the `if should_load_checkpoint: `block, which is also entered on a normal resume (resume_checkpoint_exists), not just the finetune-from-pretrained path. 

On resume, `load_checkpoint` restores the optimizer state and HDO's load hooks deliberately preserve the fp32 master to avoid precision loss — so I think force-syncing BF16 → fp32 there might overwrite the higher-precision master that was just restored. 
Could you double-check whether that path is a concern?

**Comment by RayenTian** [comment]
Thanks for tracking this down and the really clear root-cause writeup, @pengdurice 
 Would you please rebase the branch, and then we can trigger CI.

File: nemo_rl/models/megatron/setup.py (hunk 1 above)
Comment (pengdurice) [reply nemo_rl/models/megatron/setup.py]: Hi, yeah, that indeed is a concern, and is fixed by only calling it when optimizer is not none and it is fine-tune. LMK if that makes sense, thank you;-)

**Comment by pengdurice** [comment]
> Thanks for tracking this down and the really clear root-cause writeup, @pengdurice Would you please rebase the branch, and then we can trigger CI.

just rebased, thank you!

**Comment by pengdurice** [comment]
> > Thanks for tracking this down and the really clear root-cause writeup, @pengdurice Would you please rebase the branch, and then we can trigger CI.
> 
> just rebased, thank you!

@RayenTian , would you mind give it a CICD run when you get a chance? thank you!

**Comment by RayenTian** [comment]
/ok to test [9611687](https://github.com/NVIDIA-NeMo/RL/pull/2372/commits/961168720f78b279aec4a494200f498ce87c1a8c)

**Comment by pengdurice** [comment]
> /ok to test [9611687](https://github.com/NVIDIA-NeMo/RL/pull/2372/commits/961168720f78b279aec4a494200f498ce87c1a8c)

the lint check failed and I just fixed it, would you mind run cicd again? thank you! @RayenTian

**Comment by RayenTian** [comment]
/ok to test [7c12b8c](https://github.com/NVIDIA-NeMo/RL/pull/2372/commits/7c12b8c1f5f07a4ee6bc5e7ea5132e4b10f2ea78)

**Comment by pengdurice** [comment]
> /ok to test [7c12b8c](https://github.com/NVIDIA-NeMo/RL/pull/2372/commits/7c12b8c1f5f07a4ee6bc5e7ea5132e4b10f2ea78)

@RayenTian , thank you for triggering the CI. there is one job failed, but I believe that's irrelevant to my code change and it is due to the flaky test itself, would you mind take a look? thank you!

**Comment by pengdurice** [comment]
> > /ok to test [7c12b8c](https://github.com/NVIDIA-NeMo/RL/pull/2372/commits/7c12b8c1f5f07a4ee6bc5e7ea5132e4b10f2ea78)
> 
> @RayenTian , thank you for triggering the CI. there is one job failed, but I believe that's irrelevant to my code change and it is due to the flaky test itself, would you mind take a look? thank you!

@RayenTian , friendly reminder of this thread;-) Or let me know  if I can do anything!

**Comment by RayenTian** [comment]
Hi, @pengdurice. Sorry for late reply. It seems a flaky failed. Re-trigger CI now.

**Comment by pengdurice** [comment]
> Hi, @pengdurice. Sorry for late reply. It seems a flaky failed. Re-trigger CI now.

@RayenTian , thank you! now there is the test coverage issue, however it doesn't seem to come from my PR change: it is likely introduced by other code changed then merged to this branch. Would you mind take a look? Thank you!

**Comment by RayenTian** [comment]
@terrykong , would you please help take a look to this PR and merge it if it's okay?

**Review by terrykong — APPROVED** [review]
